Feature/141 automatic changelog

uluizeduardo opened the pull request:
This pull request introduces a new feature in BuildCLI for automatically generating a structured changelog based on the get commit history. The changelog generator parses commit messages, categorizes changes using predefined tags (e.g., feat, fix, docs, refactor), and supports multiple output formats such as Markdown, HTML, and JSON.

omatheusmesmo reviewed (changes requested):
@uluizeduardo , great job, it'll improve user's experience a lot. :1st_place_medal: 

I had a minor problem testing, see the log below, but it's seem its just a little fix in formatOutputFile() method. Could you please adjust that? I just run `buildcli changelog` in my BuildCLI local repository.

For your knowledge, I'm using a Debian distro. 

```
omatheusmesmo@omatheusmesmo:/media/omatheusmesmo/FileShare/BuildCLI$ java -jar ./target/buildcli.jar changelog
,-----.          ,--.,--.   ,--. ,-----.,--.   ,--.
|  |) /_ ,--.,--.`--'|  | ,-|  |'  .--./|  |   |  |
|  .-.  \|  ||  |,--.|  |' .-. ||  |    |  |   |  |       Built by the community, for the community
|  '--' /'  ''  '|  ||  |\ `-' |'  '--'\|  '--.|  |
`------'  `----' `--'`--' `---'  `-----'`-----'`--'

Failed to generate changelog.
java.lang.NullPointerException: Cannot invoke "String.contains(java.lang.CharSequence)" because "fileName" is null
        at org.buildcli.commands.ChangelogCommand.formatOutputFile(ChangelogCommand.java:111)
        at org.buildcli.commands.ChangelogCommand.run(ChangelogCommand.java:99)
        at picocli.CommandLine.executeUserObject(CommandLine.java:2030)
        at picocli.CommandLine.access$1500(CommandLine.java:148)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2465)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2457)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2419)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2277)
        at picocli.CommandLine$RunLast.execute(CommandLine.java:2421)
        at picocli.CommandLine.execute(CommandLine.java:2174)
        at org.buildcli.CommandLineRunner.main(CommandLineRunner.java:14)
java.lang.RuntimeException: java.lang.NullPointerException: Cannot invoke "String.contains(java.lang.CharSequence)" because "fileName" is null
        at org.buildcli.commands.ChangelogCommand.run(ChangelogCommand.java:106)
        at picocli.CommandLine.executeUserObject(CommandLine.java:2030)
        at picocli.CommandLine.access$1500(CommandLine.java:148)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2465)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2457)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2419)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2277)
        at picocli.CommandLine$RunLast.execute(CommandLine.java:2421)
        at picocli.CommandLine.execute(CommandLine.java:2174)
        at org.buildcli.CommandLineRunner.main(CommandLineRunner.java:14)
Caused by: java.lang.NullPointerException: Cannot invoke "String.contains(java.lang.CharSequence)" because "fileName" is null
        at org.buildcli.commands.ChangelogCommand.formatOutputFile(ChangelogCommand.java:111)
        at org.buildcli.commands.ChangelogCommand.run(ChangelogCommand.java:99)
        ... 9 more

ATTENTION: Your BuildCLI is outdated!
```

Fix suggestion:

```java

protected String formatOutputFile(String fileName, String format) {
        if (fileName == null) {
            return "CHANGELOG" + FileTypes.fromExtension(format);
        }

        String outputFileName = fileName.contains(".") ?
                fileName.substring(0, fileName.lastIndexOf('.')) :
                fileName;

        return Path.of(outputFileName + FileTypes.fromExtension(format)).toString();
    }
    
```
    
I also noted that the generated changelog with this fix is a little differente from the keep a change log suggestion, am I correct?
   
   keepachangelog example 
```
    
   # Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

- v1.1 Brazilian Portuguese translation.
- v1.1 German Translation
```
Generated by your function
```
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [v0.0.3] = 2025-02-21

### Added
- - feat (changelog): Add usage instructions for changelog generation in BuildCLI documentation
- - feat : auto-assign worklow - fixes #107
- - feat : auto-assign worklow - fixes #107
...
```

uluizeduardo commented:
I made the following adjustments based on your feedback:

- Update the `formatOutputFile` method to handle cases where the output file name is `null` or empty.
- Added a unit test to validate the formatOutputFile behavior.
- Adjusted the changelog format according to the discussed improvements:
  - Removed the duplicate dash so that each item appears as `- feat(changelog): ...` (with only one `-`).
  - Replaced `=` with `-` in the version line, making it  `## [v0.0.3] - 2025-02-21`.
  
Regarding the difference between our changelog and keep a changelog, the key distinction is that we are displaying full commit message (e.g., "feat(changelog): ..."),
while the keep a changelog example usually contains manually summarized descriptions. This choice was made to ensure clarity and traceability of changes, but was can refine it further if needed.

With these changes, the generated changelog should be much closer to the keep a Changelog style. If there are still any discrepancies or formatting preferences, we can adjust as needed!

Bellow, I'll leave the command to generate a changelog and the image of the generated changelog

`buildcli changelog -v 0.0.1 --format markdown --output CHANGELOG.md --include feat,fix,docs,chore,refactor`

![image](https://github.com/user-attachments/assets/551a75bf-d7b1-4482-a5ad-597eea964d7c)

omatheusmesmo commented:
Well done, @uluizeduardo , really a great job! Congrats on your firts PR! :clap:
